posts/fantastic-simprocs.md

```diff
@@ -0,0 +1,198 @@
+---
+author: 'Yaël Dillies, Paul Lezeau'
+category: 'Metaprogramming'
+date: 2025-03-29 12:33:00 UTC+00:00
+description: 'Brief introduction to simprocs and what they are useful for'
+has_math: true
+link: ''
+slug: fantastic-simprocs
+tags: 'simp, simproc, meta'
+title: Fantastic Simprocs and Where to Find Them.
+type: text
+---
+
+Lean core added support for custom simplification procedures, aka *simprocs*.
+This blog post is the first in a series of two aimed at explaining what a simproc is, what kind of problems can be solved with simprocs, and what tools we have to write them.
+
+<!-- TEASER_END -->
+
+This post describes purely informally what simprocs are and do.
+The second post will be a walkthrough to writing a simple simproc in three different ways.
+
+To understand what a simproc is and how it works, we will first explain how `simp` works.
+Then we will give some examples and non-examples of simprocs.
+
+# How simp works
+
+`simp` is made of two components.
+
+The first component is **rewriting rules**.
+Almost all rewriting rules are lemmas that prove an equality and are tagged with `@[simp]` in eg Lean or Mathlib.
+A rewriting rule is characterised by its *left hand side* and *right hand side*.
+Eg for a lemma of the form `LHS = RHS`, this is `LHS` and `RHS` respectively.
+
+The second one is the **`simp` tactic**, which we will refer to simply as `simp`.
+When run on a goal `⊢ e`, `simp` iteratively looks for a subexpression of `e` that matches the left hand side of some rewriting rule, and replaces that subexpression with the right hand side of that rule.
+
+For example, here's the proof steps `simp` follows to close the goal `37 * (Nat.fib 0 + 0) = 0`
+```lean
+⊢ 37 * (Nat.fib 0 + 0) = 0
+⊢ 37 * (0 + 0) = 0
+⊢ 37 * 0 = 0
+⊢ 0 = 0
+⊢ True
+```
+
+> The order in which `simp` traverses an expression is relatively intricate.
+> As a simple approximation, simplification is performed from the inside-out.
+> See the next blog post for full details on the algorithm.
+
+> If you write `set_option trace.Meta.Tactic.simp true in example : MyGoal := by simp`, you will see the list of simplification steps `simp` performs on `Goal`.
```

Comment:
Do you definitely want all of this as two block quotes? I don't think they're significant enough for that!

Reply:
Oh I thought that was pretty nice. Imagining myself reading this blog post as a beginner, these block quotes are answering the exact two questions I think I would be asking myself.

Or do you mean that there's better formatting than a block quote? I was using that as an aparté

Reply:
I think they can just be text as usual: block quotes in this way are usually used to indicate key takeaways from the overall post; these are useful points but I don't think they need highlighting in this way.

Reply:
But this is predicated on the way block quotes are displayed in the blog; perhaps the disagreement here is just that we're picturing different things!